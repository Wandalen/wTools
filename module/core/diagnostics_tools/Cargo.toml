[package]
name = "diagnostics_tools"
version = "0.8.0"
edition = "2021"
authors = [
  "Kostiantyn Wandalen <wandalen@obox.systems>",
  "Dmytro Kryvoruchko <dmytro.kr@obox.systems>",
]
license = "MIT"
readme = "Readme.md"
documentation = "https://docs.rs/diagnostics_tools"
repository = "https://github.com/Wandalen/wTools/tree/master/module/core/diagnostics_tools"
homepage = "https://github.com/Wandalen/wTools/tree/master/module/core/diagnostics_tools"
description = """
Diagnostics tools.
"""
categories = [ "algorithms", "development-tools" ]
keywords = [ "fundamental", "general-purpose" ]

[lints]
workspace = true

[package.metadata.docs.rs]
features = [ "full" ]
all-features = false



[features]
default = [
  "enabled",
  "diagnostics_runtime_assertions",
  "diagnostics_compiletime_assertions",
  "diagnostics_memory_layout",
]
full = [
  "enabled",
  "diagnostics_runtime_assertions",
  "diagnostics_compiletime_assertions",
  "diagnostics_memory_layout",
]
no_std = []
use_alloc = [ "no_std" ]
enabled = []

diagnostics_runtime_assertions = [ "pretty_assertions" ] # run-time assertions
diagnostics_compiletime_assertions = [] # compile-time assertions
diagnostics_memory_layout = [] #

[dependencies]
pretty_assertions = { version = "~1.4.0", optional = true }

[dev-dependencies]
test_tools = { workspace = true }
