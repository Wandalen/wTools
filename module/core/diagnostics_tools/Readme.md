<!-- {{# generate.module_header{} #}} -->

# Module :: diagnostics_tools
<!--{ generate.module_header.start() }-->
 [![experimental](https://raster.shields.io/static/v1?label=&message=experimental&color=orange)](https://github.com/emersion/stability-badges#experimental) [![rust-status](https://github.com/Wandalen/wTools/actions/workflows/module_diagnostics_tools_push.yml/badge.svg)](https://github.com/Wandalen/wTools/actions/workflows/module_diagnostics_tools_push.yml) [![docs.rs](https://img.shields.io/docsrs/diagnostics_tools?color=e3e8f0&logo=docs.rs)](https://docs.rs/diagnostics_tools) [![Open in Gitpod](https://raster.shields.io/static/v1?label=try&message=online&color=eee&logo=gitpod&logoColor=eee)](https://gitpod.io/#RUN_PATH=.,SAMPLE_FILE=module%2Fcore%2Fdiagnostics_tools%2Fexamples%2Fdiagnostics_tools_trivial.rs,RUN_POSTFIX=--example%20diagnostics_tools_trivial/https://github.com/Wandalen/wTools) [![discord](https://img.shields.io/discord/872391416519737405?color=eee&logo=discord&logoColor=eee&label=ask)](https://discord.gg/m3YfbXpUUY)
<!--{ generate.module_header.end }-->

Diagnostics tools.

### Basic use-case

<!-- {{# generate.module{} #}} -->

```rust
#[ test ]
#[ should_panic ]
fn a_id_panic_test()
{
  a_id!( 1, 2 );
  /*
    print :
    ...

thread 'a_id_panic_test' panicked at 'assertion failed: `(left == right)`

Diff < left / right > :
<1
>2
...
  */
}
```
<!-- zzz : qqq : add --> <!-- aaa : added -->

### To add to your project

```sh
cargo add diagnostics_tools
```

### Try out from the repository

```sh
git clone https://github.com/Wandalen/wTools
cd wTools
cd examples/diagnostics_tools_trivial
cargo run
```
