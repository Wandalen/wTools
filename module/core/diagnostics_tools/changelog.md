* [2025-07-26 13:33 UTC] Resolved stuck doctest by using `std::panic::catch_unwind` due to `should_panic` not working with `include_str!`.
* [2025-07-26 13:37 UTC] Refactored `trybuild` setup to be robust and idiomatic, consolidating compile-time assertion tests.
*   Applied `rustfmt` to the crate.
*   Fixed clippy warnings and missing documentation.