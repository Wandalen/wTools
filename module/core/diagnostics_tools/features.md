# Features and Configuration

This document describes the feature flags and configuration options available in `diagnostics_tools`.

## Default Features

By default, the crate enables these features:

```toml
[dependencies]
diagnostics_tools = "0.11"  # Includes: enabled, runtime, compiletime, memory_layout
```

This gives you access to all assertion types:
- Runtime assertions (`a_*` macros)
- Compile-time assertions (`cta_*` macros)  
- Memory layout validation (`cta_type_*`, `cta_ptr_*`, `cta_mem_*`)

## Available Feature Flags

### Core Features

#### `enabled` *(default)*
Master switch that enables the crate functionality. Without this, all macros become no-ops.

```toml
[dependencies]
diagnostics_tools = { version = "0.11", features = ["enabled"] }
```

#### `full` 
Enables all features - equivalent to enabling all individual feature flags.

```toml
[dependencies]
diagnostics_tools = { version = "0.11", features = ["full"] }
```

### Functionality Features

#### `diagnostics_runtime_assertions` *(default)*
Enables runtime assertion macros:
- `a_true!`, `a_false!`
- `a_id!`, `a_not_id!`  
- `a_dbg_true!`, `a_dbg_false!`, `a_dbg_id!`, `a_dbg_not_id!`

```toml
[dependencies]
diagnostics_tools = { version = "0.11", features = ["enabled", "diagnostics_runtime_assertions"] }
```

#### `diagnostics_compiletime_assertions` *(default)*
Enables compile-time assertion macros:
- `cta_true!`

```toml
[dependencies]
diagnostics_tools = { version = "0.11", features = ["enabled", "diagnostics_compiletime_assertions"] }
```

#### `diagnostics_memory_layout` *(default)*
Enables memory layout validation macros:
- `cta_type_same_size!`, `cta_type_same_align!`
- `cta_ptr_same_size!`, `cta_mem_same_size!`

```toml
[dependencies]
diagnostics_tools = { version = "0.11", features = ["enabled", "diagnostics_memory_layout"] }
```

### Environment Features

#### `no_std`
Enables no_std support for embedded and constrained environments.

```toml
[dependencies]
diagnostics_tools = { version = "0.11", features = ["no_std", "enabled"] }
```

When `no_std` is enabled:
- Runtime assertions still work but with limited formatting
- Compile-time assertions work exactly the same
- Memory layout validation works exactly the same

#### `use_alloc`
When using `no_std`, you can still enable heap allocation with `use_alloc`.

```toml
[dependencies]  
diagnostics_tools = { version = "0.11", features = ["no_std", "use_alloc", "enabled"] }
```

## Custom Feature Combinations

### Minimal Runtime Only
For projects that only need runtime assertions:

```toml
[dependencies]
diagnostics_tools = { 
    version = "0.11", 
    default-features = false, 
    features = ["enabled", "diagnostics_runtime_assertions"] 
}
```

### Compile-Time Only  
For projects that only need compile-time validation:

```toml
[dependencies]
diagnostics_tools = { 
    version = "0.11", 
    default-features = false, 
    features = ["enabled", "diagnostics_compiletime_assertions"] 
}
```

### Memory Layout Only
For low-level code that only needs memory validation:

```toml
[dependencies]
diagnostics_tools = { 
    version = "0.11", 
    default-features = false, 
    features = ["enabled", "diagnostics_memory_layout"] 
}
```

### Embedded/No-Std
For embedded projects:

```toml
[dependencies]
diagnostics_tools = { 
    version = "0.11", 
    default-features = false, 
    features = ["no_std", "enabled", "diagnostics_compiletime_assertions", "diagnostics_memory_layout"] 
}
```

## Conditional Compilation

You can conditionally enable features based on your build configuration:

```toml
[dependencies]
diagnostics_tools = { version = "0.11", default-features = false, features = ["enabled"] }

[dependencies.diagnostics_tools.features]
# Only include runtime assertions in debug builds
diagnostics_runtime_assertions = { optional = true }

[features]
default = []
debug_asserts = ["diagnostics_tools/diagnostics_runtime_assertions"]
```

Then use with:
```bash
# Development build with runtime assertions
cargo build --features debug_asserts

# Release build without runtime assertions  
cargo build --release
```

## Performance Impact

### Feature Impact on Binary Size

| Feature | Binary Size Impact | Runtime Impact |
|---------|-------------------|----------------|
| `diagnostics_runtime_assertions` | Medium (includes pretty_assertions) | Same as standard assertions |
| `diagnostics_compiletime_assertions` | None (compile-time only) | None |
| `diagnostics_memory_layout` | None (compile-time only) | None |
| `no_std` | Reduces size | Slightly reduced formatting |

### Recommendation by Use Case

**Testing/Development:**
```toml
diagnostics_tools = "0.11"  # Use all default features
```

**Production Libraries:**
```toml
diagnostics_tools = { 
    version = "0.11", 
    default-features = false, 
    features = ["enabled", "diagnostics_compiletime_assertions", "diagnostics_memory_layout"] 
}
```

**Embedded Systems:**
```toml
diagnostics_tools = { 
    version = "0.11", 
    default-features = false, 
    features = ["no_std", "enabled", "diagnostics_compiletime_assertions"] 
}
```

**High-Performance Applications:**
```toml
# Development
[dependencies.diagnostics_tools]
version = "0.11"

# Production (disable runtime assertions)
[dependencies.diagnostics_tools]  
version = "0.11"
default-features = false
features = ["enabled", "diagnostics_compiletime_assertions", "diagnostics_memory_layout"]
```

## Feature Interaction

Some features have dependencies on each other:

- `enabled` is required for any functionality
- `use_alloc` requires `no_std`
- All diagnostic features require `enabled`

The crate will give compile-time errors if incompatible features are selected.