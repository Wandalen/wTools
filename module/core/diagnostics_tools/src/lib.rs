#![ cfg_attr( feature = "no_std", no_std ) ]
#![ doc( html_logo_url = "https://raw.githubusercontent.com/Wandalen/wTools/master/asset/img/logo_v3_trans_square.png" ) ]
#![ doc( html_favicon_url = "https://raw.githubusercontent.com/Wandalen/wTools/alpha/asset/img/logo_v3_trans_square_icon_small_v2.ico" ) ]
#![ doc( html_root_url = "https://docs.rs/diagnostics_tools/latest/diagnostics_tools/" ) ]
#![ doc = include_str!( concat!( env!( "CARGO_MANIFEST_DIR" ), "/", "Readme.md" ) ) ]

#[ cfg( feature = "enabled" ) ]
/// Compile-time asserting.
pub mod diag;

/// Namespace with dependencies.

#[ cfg( feature = "enabled" ) ]
pub mod dependency
{
  #[ cfg( feature = "diagnostics_runtime_assertions" ) ]
  pub use ::pretty_assertions;
}

#[ doc( inline ) ]
#[ allow( unused_imports ) ]
#[ cfg( feature = "enabled" ) ]
pub use own::*;

/// Own namespace of the module.
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub mod own
{
  use super::*;
  #[ doc( inline ) ]
  pub use orphan::*;
  #[ doc( inline ) ]
  #[ allow( unused_imports ) ]
  pub use super::diag::orphan::*;
}

/// Orphan namespace of the module.
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub mod orphan
{
  use super::*;
  #[ doc( inline ) ]
  pub use exposed::*;
}

/// Exposed namespace of the module.
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub mod exposed
{
  use super::*;
  #[ doc( inline ) ]
  pub use prelude::*;
  #[ doc( inline ) ]
  #[ allow( unused_imports ) ]
  pub use super::diag::exposed::*;
}

/// Prelude to use essentials: `use my_module::prelude::*`.
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub mod prelude
{
  use super::*;
  #[ doc( inline ) ]
  #[ allow( unused_imports ) ]
  pub use super::diag::prelude::*;
}
