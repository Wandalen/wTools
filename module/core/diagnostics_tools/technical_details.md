# Technical Details

This document contains implementation details and technical information for the `diagnostics_tools` crate.

## Architecture Overview

The crate is organized into three main modules:

- **`rta`** - Runtime assertions (Runtime-Time Assertions)
- **`cta`** - Compile-time assertions (Compile-Time Assertions)  
- **`layout`** - Memory layout validation

## Module Structure

### Runtime Assertions (`rta`)

All runtime assertion macros follow the pattern `a_*` (assertion):

- `a_true!(condition)` - Assert condition is true
- `a_false!(condition)` - Assert condition is false
- `a_id!(left, right)` - Assert values are identical (equal)
- `a_not_id!(left, right)` - Assert values are not identical

Debug variants (`a_dbg_*`) print values even when assertions pass:

- `a_dbg_true!(condition)`
- `a_dbg_false!(condition)`
- `a_dbg_id!(left, right)`
- `a_dbg_not_id!(left, right)`

### Compile-Time Assertions (`cta`)

- `cta_true!(condition)` - Compile-time boolean check using `cfg` conditions

### Memory Layout Validation (`layout`)

- `cta_type_same_size!(Type1, Type2)` - Verify types have same size
- `cta_type_same_align!(Type1, Type2)` - Verify types have same alignment
- `cta_ptr_same_size!(ptr1, ptr2)` - Verify pointers have same size
- `cta_mem_same_size!(val1, val2)` - Verify values have same memory size

## Implementation Details

### Error Message Enhancement

The crate uses `pretty_assertions` internally to provide:
- Colored diff output
- Structured comparison formatting
- Better visual distinction between expected and actual values

### Compile-Time Validation

Compile-time assertions use Rust's `compile_error!` macro combined with `cfg` attributes to validate conditions during compilation.

### Memory Layout Checks

Memory layout assertions use:
- `core::mem::size_of::<T>()` for size validation
- `core::mem::align_of::<T>()` for alignment validation
- Array length tricks to force compile-time evaluation

## Feature Flags

The crate supports several feature flags for conditional compilation:

- `enabled` - Master switch for all functionality (default)
- `diagnostics_runtime_assertions` - Runtime assertion macros (default)
- `diagnostics_compiletime_assertions` - Compile-time assertion macros (default) 
- `diagnostics_memory_layout` - Memory layout validation macros (default)
- `no_std` - Support for no_std environments
- `full` - Enable all features

## Performance Considerations

### Runtime Overhead

- Runtime assertions have the same overhead as standard `assert!` macros
- Debug variants have additional overhead for value formatting
- All assertions are removed in release builds unless explicitly enabled

### Compile-Time Impact

- Compile-time assertions have zero runtime overhead
- They may slightly increase compilation time due to additional checking
- Memory layout assertions are resolved entirely at compile time

## Namespace Organization

The crate uses a hierarchical namespace structure:

```
diagnostics_tools/
├── own/       - Direct exports
├── orphan/    - Re-exports from submodules  
├── exposed/   - Extended API surface
└── prelude/   - Common imports
```

## Integration with Testing Frameworks

The runtime assertions integrate seamlessly with:
- Built-in Rust test framework (`#[test]`)
- Custom test harnesses
- Benchmark frameworks

## Error Handling Philosophy

The crate follows Rust's philosophy of "fail fast":
- Runtime assertions panic on failure (like standard assertions)
- Compile-time assertions prevent compilation on failure
- Clear, actionable error messages help identify root causes quickly

## Cross-Platform Compatibility

- Full support for all Rust-supported platforms
- `no_std` compatibility for embedded systems
- Consistent behavior across different architectures