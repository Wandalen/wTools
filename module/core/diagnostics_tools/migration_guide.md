# Migration Guide

This guide helps you migrate from standard Rust assertions to `diagnostics_tools` for better debugging experience.

## Quick Migration Table

| Standard Rust | Diagnostics Tools | Notes |
|---------------|-------------------|-------|
| `assert!(condition)` | `a_true!(condition)` | Same behavior, better error context |
| `assert!(!condition)` | `a_false!(condition)` | More explicit intent |
| `assert_eq!(a, b)` | `a_id!(a, b)` | Colored diff output |
| `assert_ne!(a, b)` | `a_not_id!(a, b)` | Colored diff output |
| `debug_assert!(condition)` | `a_dbg_true!(condition)` | Always prints values |
| `debug_assert_eq!(a, b)` | `a_dbg_id!(a, b)` | Always prints values |

## Step-by-Step Migration

### 1. Add Dependency

Update your `Cargo.toml`:

```toml
[dependencies]
# Add this line:
diagnostics_tools = "0.11"
```

### 2. Import the Prelude

Add to your source files:

```rust
// At the top of your file:
use diagnostics_tools::*;
```

Or more specifically:
```rust
use diagnostics_tools::{ a_true, a_false, a_id, a_not_id };
```

### 3. Replace Assertions Gradually

**Before:**
```rust
fn test_my_function() {
    let result = my_function();
    assert_eq!(result.len(), 3);
    assert!(result.contains("hello"));
    assert_ne!(result[0], "");
}
```

**After:**
```rust
fn test_my_function() {
    let result = my_function();
    a_id!(result.len(), 3);           // Better diff on failure
    a_true!(result.contains("hello")); // Better error context  
    a_not_id!(result[0], "");         // Better diff on failure
}
```

## Advanced Migration Scenarios

### Testing Complex Data Structures

**Before:**
```rust
#[test]
fn test_user_data() {
    let user = create_user();
    assert_eq!(user.name, "John");
    assert_eq!(user.age, 30);
    assert_eq!(user.emails.len(), 2);
}
```

**After:**
```rust
#[test]
fn test_user_data() {
    let user = create_user();
    
    // Get beautiful structured diffs for complex comparisons:
    a_id!(user, User {
        name: "John".to_string(),
        age: 30,
        emails: vec!["john@example.com".to_string(), "j@example.com".to_string()],
    });
}
```

### Adding Compile-Time Checks

**Before:**
```rust
// No equivalent - this was impossible with standard assertions
```

**After:**
```rust
// Validate assumptions at compile time:
cta_true!(cfg(feature = "serde"));
cta_type_same_size!(u32, i32);
cta_type_same_align!(u64, f64);
```

### Development vs Production

**Before:**
```rust
fn validate_input(data: &[u8]) {
    debug_assert!(data.len() > 0);
    debug_assert!(data.len() < 1024);
}
```

**After:**
```rust
fn validate_input(data: &[u8]) {
    // Debug variants show values even on success during development:
    a_dbg_true!(data.len() > 0);
    a_dbg_true!(data.len() < 1024);
    
    // Or use regular variants that only show output on failure:
    a_true!(data.len() > 0);
    a_true!(data.len() < 1024);
}
```

## Coexistence Strategy

You dont need to migrate everything at once. The crates work together:

```rust
use diagnostics_tools::*;

fn mixed_assertions() {
    // Keep existing assertions:
    assert!(some_condition);
    
    // Add enhanced ones where helpful:
    a_id!(complex_struct_a, complex_struct_b); // Better for complex comparisons
    
    // Use compile-time checks for new assumptions:
    cta_true!(cfg(target_pointer_width = "64"));
}
```

## Common Migration Patterns

### 1. Test Suites

Focus on test files first - this is where better error messages provide the most value:

```rust
// tests/integration_test.rs
use diagnostics_tools::*;

#[test]
fn api_response_format() {
    let response = call_api();
    
    // Much clearer when JSON structures differ:
    a_id!(response, expected_json_structure());
}
```

### 2. Development Utilities

Use debug variants during active development:

```rust
fn debug_data_processing(input: &Data) -> ProcessedData {
    let result = process_data(input);
    
    // Shows values even when assertions pass - helpful during development:
    a_dbg_id!(result.status, Status::Success);
    a_dbg_true!(result.items.len() > 0);
    
    result
}
```

### 3. Library Boundaries

Add compile-time validation for public APIs:

```rust
pub fn new_public_api<T>() -> T 
where 
    T: Default + Clone + Send,
{
    // Validate assumptions about T at compile time:
    cta_type_same_size!(T, T);  // Sanity check
    
    // Runtime validation with better errors:
    let result = T::default();
    a_true!(std::mem::size_of::<T>() > 0);
    
    result
}
```

## Tips for Smooth Migration

1. **Start with Tests**: Migrate test assertions first - you'll see immediate benefits
2. **Use Debug Variants During Development**: They provide extra visibility
3. **Add Compile-Time Checks Gradually**: Look for assumptions that could be validated earlier
4. **Focus on Complex Comparisons**: The biggest wins come from comparing structs, vectors, and other complex data
5. **Keep It Mixed**: You dont need to replace every assertion - focus on where enhanced messages help most

## Rollback Strategy

If you need to rollback temporarily, simply:

1. Remove the `use diagnostics_tools::*;` import
2. Use find-replace to convert back:
   - `a_true!` → `assert!`
   - `a_id!` → `assert_eq!`  
   - `a_not_id!` → `assert_ne!`
   - Remove any compile-time assertions (they have no standard equivalent)

The migration is designed to be low-risk and reversible.