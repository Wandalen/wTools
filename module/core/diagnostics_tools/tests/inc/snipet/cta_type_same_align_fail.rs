use test_tools::diagnostics_tools::*;

fn main()
{
  #[ repr( align( 128 ) ) ]
  struct Int( i16 );
  cta_type_same_align!( Int, i16 );
}
