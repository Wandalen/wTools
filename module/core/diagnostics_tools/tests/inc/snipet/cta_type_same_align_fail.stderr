error[E0308]: mismatched types
 --> tests/inc/snipet/cta_type_same_align_fail.rs:7:3
  |
7 |   cta_type_same_align!( Int, i16 );
  |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   |
  |   expected an array with a size of 128, found one with a size of 2
  |   expected due to this
  |
  = note: this error originates in the macro `cta_type_same_align` (in Nightly builds, run with -Z macro-backtrace for more info)
